### PR TITLE
Abort on std::logic_error in debug builds

### PR DESCRIPTION
**Pull request by akuzm**
This helps notice them in tests, same as assert() and LOGICAL_ERROR.

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

File: src/Server/TCPHandler.cpp [hunk 1]
```diff
@@ -313,6 +313,15 @@ void TCPHandler::runImpl()
...
             exception.emplace(Exception::CreateFromPocoTag{}, e);
         }
+        catch (const std::logic_error & e)
+        {
+            state.io.onException();
+            exception.emplace(Exception::CreateFromSTDTag{}, e);
+            // Server should die on std logic errors in debug, like with assert()
+            // or ErrorCodes::LOGICAL_ERROR. This helps catch these errors in
+            // tests.
+            assert(false);
```
Comment (nikitamikhaylov) [inline src/Server/TCPHandler.cpp:323]: `std::terminate`?

**Review by nikitamikhaylov — APPROVED** [review]

File: src/Server/TCPHandler.cpp (hunk 1 above)
Comment (akuzm) [reply src/Server/TCPHandler.cpp:323]: I only want it in a debug build, so assert is more convenient.

**Comment by akuzm** [comment]
```
SELECT toRelativeDayNum(( SELECT dictGetUInt8() ) AS a), NULL, ( SELECT * FROM hdfs(['B:q']) ) AS st, []; 
```

sql_fuzzy.pl finds an assertion right away

**Comment by akuzm** [comment]
https://clickhouse-test-reports.s3.yandex.net/12522/a4a23673cc640ab77efc9315a919519a9e49ed89/functional_stateless_tests_(debug)/test_run.txt.out.log

`Logical error: 'Sort order of blocks violated for column ``, left: UInt64_255, right: UInt64_0.'.`

Looks like it's not related to the changes.

**Pull request merged**